reactive/rerunner.go

```diff
@@ -225,24 +174,46 @@ func Cache(ctx context.Context, key interface{}, f ComputeFunc) (interface{}, er
 	}
 
 	cache := ctx.Value(cacheKey{}).(*cache)
-	computation := ctx.Value(computationKey{}).(*computation)
+	parent := ctx.Value(computationKey{}).(*computation)
 
-	cache.locker.Lock(key)
-	defer cache.locker.Unlock(key)
-
-	if child := cache.get(key); child != nil {
-		child.node.addOut(&computation.node)
-		return child.value, nil
+	cache.mu.Lock()
+	entry, ok := cache.entries[key]
+	if !ok {
+		// First cache call for this key.
+		entry = &cacheEntry{
+			done: make(chan struct{}),
+		}
+		cache.entries[key] = entry
+	}
+	cache.mu.Unlock()
+
+	if ok {
```

Comment:
non-blocking

To reduce some of the logic here, should we invert the check here and use the same logic for lookup/return?
```
if !ok {
  entry.computation, entry.err = run(ctx, f) // close entry.Done in this func
}
select {
  case <-ctx.Done():
     ...
  case <-entry.done:
}
if entry.computation == nil {
  ...
}
if entry.err != nil {
 ...
}
//add out
return value
```